src/C-interface/ellpack/bml_transpose_ellpack_typed.c

```diff
@@ -204,5 +217,118 @@ void TYPED_FUNC(
+    REAL_T *csrValA = (REAL_T *) A->csrVal;
+
+    /* temporary arrays to hold compressed sparse column (CSC) values */
+    int *cscRowInd = NULL;
+    int *cscColPtr = NULL;
+    REAL_T *cscVal = NULL;
+
+    cusparseStatus_t status = CUSPARSE_STATUS_SUCCESS;
+    cudaDataType valType = BML_CUSPARSE_T;
+
+    // CUSPARSE APIs
+    cusparseHandle_t handle = NULL;
+    void *dBuffer = NULL;
+    size_t bufferSize = 0;
+
+    // convert ellpack to cucsr
+    TYPED_FUNC(bml_ellpack2cucsr_ellpack) (A);
+
+    // Create cusparse matrix A and B in CSR format
+    // Note: The following update is not necessary since the ellpack2cucsr
+    // routine updates the csr rowpointers on host and device
+#pragma omp target update from(csrRowPtrA[:N+1])
+    int nnzA = csrRowPtrA[N];
+
+    // Allocate memory for result arrays
+    cscVal =
+        (REAL_T *) omp_target_alloc(sizeof(REAL_T) * nnzA,
+                                    omp_get_default_device());
+    cscRowInd =
+        (int *) omp_target_alloc(sizeof(int) * nnzA,
+                                 omp_get_default_device());
+    cscColPtr =
+        (int *) omp_target_alloc(sizeof(int) * (N + 1),
+                                 omp_get_default_device());
+
+    BML_CHECK_CUSPARSE(cusparseCreate(&handle));
+#pragma omp target data use_device_ptr(csrRowPtrA,csrColIndA,csrValA)
+    {
+        // Get storage buffer size
+        BML_CHECK_CUSPARSE(cusparseCsr2cscEx2_bufferSize
+                           (handle, N, N, nnzA, csrValA, csrRowPtrA,
+                            csrColIndA, cscVal, cscColPtr, cscRowInd, valType,
+                            CUSPARSE_ACTION_NUMERIC, CUSPARSE_INDEX_BASE_ZERO,
+                            CUSPARSE_CSR2CSC_ALG1, &bufferSize));
+
+        // Allocate buffer and perform transpose operation
+        dBuffer =
+            (char *) omp_target_alloc(bufferSize, omp_get_default_device());
+
+        BML_CHECK_CUSPARSE(cusparseCsr2cscEx2
+                           (handle, N, N, nnzA, csrValA, csrRowPtrA,
+                            csrColIndA, cscVal, cscColPtr, cscRowInd, valType,
+                            CUSPARSE_ACTION_NUMERIC, CUSPARSE_INDEX_BASE_ZERO,
+                            CUSPARSE_CSR2CSC_ALG1, dBuffer));
+
+        /* Update matA with new result. Note that only device arrays are updated. */
+        omp_target_memcpy(csrRowPtrA, cscColPtr, (N + 1) * sizeof(int),
+                          0, 0, omp_get_default_device(),
+                          omp_get_default_device());
+        omp_target_memcpy(csrColIndA, cscRowInd, nnzA * sizeof(int), 0,
+                          0, omp_get_default_device(),
+                          omp_get_default_device());
+        omp_target_memcpy(csrValA, cscVal, nnzA * sizeof(REAL_T), 0, 0,
+                          omp_get_default_device(), omp_get_default_device());
+
+        // deallocate storage buffer
+        omp_target_free(dBuffer, omp_get_default_device());
+    }
+
+/*
+// DEBUG:
```

Comment:
Please remove commented out code.

Reply:
@nicolasbock comment has been removed. BTW there are other commented out code that was not introduced in this PR. I left those intact. Thanks.